flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/filesystem/LimitableBulkFormat.java

```diff
@@ -105,8 +111,16 @@ private boolean reachLimit() {
                 return null;
             }
 
-            RecordIterator<T> batch = reader.readBatch();
-            return batch == null ? null : new LimitableIterator(batch);
+            try {
+                RecordIterator<T> batch = reader.readBatch();
+                return batch == null ? null : new LimitableIterator(batch);
+            } catch (Exception e) {
```

Comment:
nit: the java.lang.IllegalStateException(reported in the issue) is a RuntimeException, not an IOException, so it doesn't match the method signature.

Reply:
I think we can ignore other exceptions too. After all, it has read enough records.

Reply:
This change is a no-op in a sequential threading model. If `reachLimit()` returns true, then the first `if` in `readBatch` already returns `null`. In all other cases, the exception is rethrown.

So I'm assuming you are actually guarding against some concurrent modification in another thread. If so, then I'd rather fix the threading model. This class already uses two different mechanisms of dealing with concurrency (`synchronized` and `AtomicLong`) and you now add optimistic invocation with exception handling as a third. This is a hard to reason. You probably should settle with using a `Lock` and use it all the way.

Btw I don't understand many parts of this class: 
- What's the purpose of `globalNumberRead`? The bulk format instance will be replicated per subtask, so why would you need it?
- Do you ensure that only one subtask is executed or how do you consolidate the different lists of the different subtask? `LIMIT N` would produce a list of `N` records in each subtask, right?

Reply:
First,
- The `synchronized` is for the creation of the counter.
- The counter is for the limit number.
They are for different purposes.

LIMIT N would produce a list of N records totally. But the source can produce more. There are limiters in the downstream operator.
So the `globalNumberRead` is just for the total limit, as soon as someone has reached its limit, the source can be ended.